src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreMvcCoreBuilderExtensions.cs

```diff
@@ -97,51 +97,23 @@ internal static void AddAuthorizationServices(IServiceCollection services)
         }
 
         /// <summary>
-        /// Register the specified <paramref name="controllerTypes"/> as services and as a source for controller
-        /// discovery.
+        /// Registers discovered controllers as services in the <see cref="IServiceCollection"/>.
         /// </summary>
         /// <param name="builder">The <see cref="IMvcCoreBuilder"/>.</param>
-        /// <param name="controllerTypes">A sequence of controller <see cref="Type"/>s to register.</param>
         /// <returns>The <see cref="IMvcCoreBuilder"/>.</returns>
-        public static IMvcCoreBuilder AddControllersAsServices(
-           this IMvcCoreBuilder builder,
-           params Type[] controllerTypes)
+        public static IMvcCoreBuilder AddControllersAsServices(this IMvcCoreBuilder builder)
         {
-            return builder.AddControllersAsServices(controllerTypes.AsEnumerable());
-        }
+            var feature = new ControllerFeature();
+            builder.PartManager.PopulateFeature(feature);
 
-        /// <summary>
-        /// Register the specified <paramref name="controllerTypes"/> as services and as a source for controller
-        /// discovery.
-        /// </summary>
-        /// <param name="builder">The <see cref="IMvcCoreBuilder"/>.</param>
-        /// <param name="controllerTypes">A sequence of controller <see cref="Type"/>s to register.</param>
-        /// <returns>The <see cref="IMvcCoreBuilder"/>.</returns>
-        public static IMvcCoreBuilder AddControllersAsServices(
-           this IMvcCoreBuilder builder,
-           IEnumerable<Type> controllerTypes)
-        {
-            if (builder == null)
+            foreach (var controller in feature.Controllers.Select(c => c.AsType()))
             {
-                throw new ArgumentNullException(nameof(builder));
+                builder.Services.TryAddTransient(controller, controller);
             }
 
-            ControllersAsServices.AddControllersAsServices(builder.Services, controllerTypes);
-            return builder;
-        }
+            builder.Services.Replace(ServiceDescriptor.Transient<IControllerActivator, ServiceBasedControllerActivator>());
```

Comment:
There's a small tradeoff, the service provide has to hold on to it. We generally only make 'top-level' services into singletons. In this case neither the activator nor the factory are 'top-level'.